File: sdk-go/cacache.go
package rbac

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// CACache handles loading or fetching a CA certificate from a remote server.
type CACache struct {
	baseURL string
	path    string
	client  *http.Client
}

// NewCACache creates a CACache that will fetch the CA cert from baseURL/v1/ca-cert
// and store it at path.
func NewCACache(baseURL, path string) *CACache {
	return &CACache{
		baseURL: baseURL,
		path:    path,
		client:  &http.Client{Timeout: 10 * time.Second},
	}
}

// LoadOrFetch checks if the CA cert exists at the configured path. If it does,
// it reads and returns the PEM bytes. If not, it fetches from the server's
// /v1/ca-cert endpoint, writes to the local path, and returns the PEM bytes.
func (c *CACache) LoadOrFetch(ctx context.Context) ([]byte, error) {
	// Check if local file exists
	if _, err := os.Stat(c.path); err == nil {
		// File exists, read it
		pem, err := os.ReadFile(c.path)
		if err != nil {
			return nil, fmt.Errorf("rbac: read CA cert from %q: %w", c.path, err)
		}
		if len(pem) > 0 {
			return pem, nil
		}
		// Empty file, treat as missing and refetch
	} else if !os.IsNotExist(err) {
		return nil, fmt.Errorf("rbac: stat CA cert path %q: %w", c.path, err)
	}

	// File doesn't exist or is empty, fetch from server
	return c.fetchFromServer(ctx)
}

// fetchFromServer downloads the CA cert from /v1/ca-cert with retry-once logic.
func (c *CACache) fetchFromServer(ctx context.Context) ([]byte, error) {
	var lastErr error

	for attempt := 0; attempt < 2; attempt++ {
		pem, err := c.doFetch(ctx)
		if err == nil {
			// Write to local path
			if err := os.WriteFile(c.path, pem, 0644); err != nil {
				return nil, fmt.Errorf("rbac: write CA cert to %q: %w", c.path, err)
			}
			return pem, nil
		}
		lastErr = err
	}

	return nil, fmt.Errorf("rbac: failed to fetch CA cert after 2 attempts: %w", lastErr)
}

// doFetch performs a single fetch from /v1/ca-cert.
func (c *CACache) doFetch(ctx context.Context) ([]byte, error) {
	url := c.baseURL + "/v1/ca-cert"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("rbac: build CA cert request: %w", err)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("rbac: fetch CA cert: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("rbac: fetch CA cert: server returned status %d", resp.StatusCode)
	}

	data, err := io.ReadAll(io.LimitReader(resp.Body, 64<<10)) // 64KB max
	if err != nil {
		return nil, fmt.Errorf("rbac: read CA cert response: %w", err)
	}

	if len(data) == 0 {
		return nil, fmt.Errorf("rbac: CA cert endpoint returned empty response")
	}

	return data, nil
}

File: sdk-rust/src/cacache.rs
//! Automatic CA certificate fetching and caching.

use std::fs;
use std::path::Path;

use reqwest::blocking::Client as HttpClient;

use crate::error::Error;

const MAX_CA_CERT_BYTES: u64 = 64 << 10; // 64KB

/// Handles loading or fetching a CA certificate from the server.
pub struct CACache {
    base_url: String,
    path: String,
    client: HttpClient,
}

impl CACache {
    /// Creates a new CACache that will fetch from `{base_url}/v1/ca-cert`
    /// and store at `path`.
    pub fn new(base_url: &str, path: &str) -> Self {
        let client = HttpClient::builder()
            .timeout(std::time::Duration::from_secs(10))
            .build()
            .expect("failed to build HTTP client for CA cache");

        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            path: path.to_string(),
            client,
        }
    }

    /// Checks if a CA cert exists at the configured path. If present and non-empty,
    /// reads and returns the PEM bytes. Otherwise, fetches from the server's
    /// `/v1/ca-cert` endpoint, writes to the local path, and returns the PEM bytes.
    pub fn load_or_fetch(&self) -> Result<Vec<u8>, Error> {
        // Check if local file exists
        let path = Path::new(&self.path);
        if path.exists() {
            let pem = fs::read(path).map_err(|e| {
                Error::Config(format!("read CA cert from {:?}: {}", self.path, e))
            })?;
            if !pem.is_empty() {
                return Ok(pem);
            }
            // Empty file, treat as missing and refetch
        }

        // File doesn't exist or is empty, fetch from server with retry-once
        self.fetch_from_server()
    }

    /// Downloads the CA cert from `/v1/ca-cert` with retry-once logic.
    fn fetch_from_server(&self) -> Result<Vec<u8>, Error> {
        let mut last_err = None;

        for _attempt in 0..2 {
            match self.do_fetch() {
                Ok(pem) => {
                    // Write to local path
                    fs::write(&self.path, &pem).map_err(|e| {
                        Error::Config(format!("write CA cert to {:?}: {}", self.path, e))
                    })?;
                    return Ok(pem);
                }
                Err(e) => {
                    last_err = Some(e);
                }
            }
        }

        Err(last_err.unwrap_or_else(|| {
            Error::Config("failed to fetch CA cert after 2 attempts".into())
        }))
    }

    /// Performs a single fetch from `/v1/ca-cert`.
    fn do_fetch(&self) -> Result<Vec<u8>, Error> {
        let url = format!("{}/v1/ca-cert", self.base_url);

        let resp = self
            .client
            .get(&url)
            .send()
            .map_err(|e| Error::Config(format!("fetch CA cert: {}", e)))?;

        let status = resp.status();
        if !status.is_success() {
            return Err(Error::Config(format!(
                "fetch CA cert: server returned status {}",
                status.as_u16()
            )));
        }

        let bytes = resp
            .bytes()
            .map_err(|e| Error::Config(format!("read CA cert response: {}", e)))?;

        if bytes.len() > MAX_CA_CERT_BYTES as usize {
            return Err(Error::Config(
                "CA cert response exceeds size limit".into(),
            ));
        }

        if bytes.is_empty() {
            return Err(Error::Config(
                "CA cert endpoint returned empty response".into(),
            ));
        }

        Ok(bytes.to_vec())
    }
}
